pkg/kube/krt/map.go

```diff
@@ -0,0 +1,138 @@
+	collectionName string
+	id             collectionUID
+	collection     internalCollection[T]
+	mapFunc        func(T) U
+}
+
+func (m *mapCollection[T, U]) GetKey(k string) *U {
+	if obj := m.collection.GetKey(k); obj != nil {
+		return ptr.Of(m.mapFunc(*obj))
+	}
+	return nil
+}
+
+func (m *mapCollection[T, U]) List() []U {
+	var res []U
+	for _, obj := range m.collection.List() {
+		res = append(res, m.mapFunc(obj))
+	}
+	return res
+}
+
+func (m *mapCollection[T, U]) Register(handler func(Event[U])) HandlerRegistration {
+	return m.collection.Register(func(t Event[T]) {
+		e := Event[U]{
+			Event: t.Event,
+		}
+		if t.Old != nil {
+			e.Old = ptr.Of(m.mapFunc(*t.Old))
+		}
+		if t.New != nil {
+			e.New = ptr.Of(m.mapFunc(*t.New))
+		}
+		handler(e)
+	})
+}
+
+func (m *mapCollection[T, U]) RegisterBatch(handler func([]Event[U]), runExistingState bool) HandlerRegistration {
+	return m.collection.RegisterBatch(func(t []Event[T]) {
+		var events []Event[U]
+		for _, t := range t {
+			e := Event[U]{
+				Event: t.Event,
+			}
+			if t.Old != nil {
+				e.Old = ptr.Of(m.mapFunc(*t.Old))
+			}
+			if t.New != nil {
+				e.New = ptr.Of(m.mapFunc(*t.New))
+			}
+			events = append(events, e)
+		}
+		handler(events)
+	}, runExistingState)
+}
+
+func (m *mapCollection[T, U]) Metadata() Metadata {
+	return m.collection.Metadata()
+}
+
+// nolint: unused // (not true, its to implement an interface)
+func (m *mapCollection[T, U]) augment(a any) any {
+	// not supported in this collection type
+	return a
+}
+
+// nolint: unused // (not true, its to implement an interface)
+func (m *mapCollection[T, U]) name() string { return m.collectionName }
+
+// nolint: unused // (not true, its to implement an interface)
+func (m *mapCollection[T, U]) uid() collectionUID { return m.id }
+
+// nolint: unused // (not true, its to implement an interface)
+func (m *mapCollection[T, U]) dump() CollectionDump {
+	// Dump should not be used on join; instead its preferred to enroll each individual collection. Maybe reconsider
+	// in the future if there is a need
+	return CollectionDump{}
+}
+
+// nolint: unused // (not true, its to implement an interface)
+func (m *mapCollection[T, U]) index(name string, extract func(o U) []string) kclient.RawIndexer {
+	t := func(o T) []string {
+		return extract(m.mapFunc(o))
+	}
+	return m.collection.index(name, t)
+}
+
+func (m *mapCollection[T, U]) HasSynced() bool {
+	return m.collection.HasSynced()
+}
+
+func (m *mapCollection[T, U]) WaitUntilSynced(stop <-chan struct{}) bool {
+	return m.collection.WaitUntilSynced(stop)
+}
+
+func MapCollection[T, U any](
+	collection Collection[T],
+	mapFunc func(T) U,
+	opts ...CollectionOption,
+) Collection[U] {
+	o := buildCollectionOptions(opts...)
+	if o.name == "" {
+		o.name = fmt.Sprintf("Map[%v]", ptr.TypeName[T]())
+	}
+
+	ic := collection.(internalCollection[T])
+	return &mapCollection[T, U]{
+		collectionName: o.name,
+		id:             ic.uid(), // TODO: should we use a new UID?
```

Comment:
@stevenctl @ilrudie any opinions on this? You folks have touched the core krt engine more than I have

Reply:
Pretty sure it needs it's own uid. Idk if there would be a big benefit from reusing it. 

Btw, IIUC, this is like a standard collection but without access to `kctx` during transformation and that gives us certain guarantees so we can optimize for performance, correct?

Reply:
Sorry I missed this comment. I think you're right on both counts. Specifically, MapCollection doesn't copy the underlying data in the original collection